File: spark-in-action/java/src/main/java/transform/document/CsvWithEmbdeddedJsonApp.java
package transform.document;

import static org.apache.spark.sql.functions.concat;
import static org.apache.spark.sql.functions.explode;
import static org.apache.spark.sql.functions.lit;

import java.io.Serializable;

import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;


public class CsvWithEmbdeddedJsonApp {
    private static final long serialVersionUID = 19711L;

    private final class Jsonifier implements MapFunction<Row, String> {
        private static final long serialVersionUID = 19712L;

        @Override
        public String call(Row r) throws Exception {
            System.out.println(r.mkString());
            StringBuffer sb = new StringBuffer();
            sb.append("{ \"dept\": \"");
            sb.append(r.getString(0));
            sb.append("\",");

            String s = r.getString(1);
            if (s != null) {
                s = s.trim();
                if (s.charAt(0) == '{') {
                    s = s.substring(1, s.length() - 1);
                }
            }
            sb.append(s);
            sb.append(", \"location\": \"");
            sb.append(r.getString(2));
            sb.append("\"}");

            return sb.toString();
        }
    }

    public static void main(String[] args) {
        CsvWithEmbdeddedJsonApp app = new CsvWithEmbdeddedJsonApp();
        app.start();
    }

    private void start() {
        SparkSession spark = SparkSession.builder()
                .appName("processing with csv")
                .master("local")
                .getOrCreate();

        Dataset<Row> df = spark.read()
                .option("header", "true")
                .option("delimiter", "|")
                .csv("data/ch13/misc/csv_with_embedded_json.csv");
        df.show(5, false);
        df.printSchema();

        Dataset<String> ds = df.map(
                new Jsonifier(),
                Encoders.STRING());
        ds.show(5, false);
        ds.printSchema();

        Dataset<Row> dfJson = spark.read().json(ds);
        dfJson.show(5, false);
        dfJson.printSchema();

        dfJson = dfJson
                .withColumn("emp", explode(dfJson.col("employee")));
    }
}

File: spark-in-action/scala/src/main/scala/transform/document/CsvWithEmbdeddedJsonApp.scala
package transform.document

import org.apache.spark.api.java.function.MapFunction
import org.apache.spark.sql.{Encoders, Row, SparkSession, functions => F}

class CsvWithEmbdeddedJsonApp {
  @SerialVersionUID(19712L)
  final private class Jsonifer extends MapFunction[Row, String] {
    @throws[Exception]
    override def call(r: Row) : String = {
      val sb = new StringBuffer
      sb.append("{ \"dept\": \"")
      sb.append(r.getString(0))
      sb.append("\",")

      var s = r.getString(1).toString
      if (s != null) {
        s = s.trim
        if (s.charAt(0) == '{')
          s = s.substring(1, s.length - 1)
      }

      sb.append(s)
      sb.append(", \"location\": \"")
      sb.append(r.getString(2))
      sb.append("\"}")
      sb.toString
    }
  }

  def start(): Unit = {
    // Creates a session on a local master
    val spark = SparkSession.builder
      .appName("Processing embedded json data")
      .master("local[*]")
      .getOrCreate

    spark.sparkContext.setLogLevel("ERROR")

    val df = spark.read
      .option("header", "true")
      .option("delimiter", "|")
      .csv("data/ch13/misc/csv_with_embedded_json.csv")

    df.show(5, false)
    df.printSchema

    val ds = df.map(new Jsonifer, Encoders.STRING)
    ds.show(5, false)
    ds.printSchema

    var dfJson = spark.read.json(ds)
    dfJson.show(5, false)
    dfJson.printSchema

    dfJson = dfJson
      .withColumn("emp", F.explode(F.col("employee")))
      .drop("employee")

    dfJson.show(5, false)
    dfJson.printSchema

    dfJson = dfJson
      .withColumn("emp_name",
        F.concat(F.col("emp.name.firstName"), F.lit(" "), F.col("emp.name.lastName")))
      .withColumn("emp_address",
        F.concat(F.col("emp.address.street"), F.lit(" "), F.col("emp.address.unit")))
      .drop("emp")

    dfJson.show(5, false)
    dfJson.printSchema

    spark.stop
  }
}

object CsvWithEmbeddedJsonApplication {
  def main(args: Array[String]): Unit = {
    val app = new CsvWithEmbdeddedJsonApp
    app.start()
  }
}